File: algo_05_dynamic_plan/L64-m.cpp
//
// Created by LiHeng on 2020/4/22.
//
#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
    int minPathSum(vector<vector<int>>& grid) {
        int m= grid.size();
        int n= grid[0].size();
        vector<vector<int>> res(m,vector<int>(n,grid[0][0]));

        for(int i=1; i<n; i++)
            res[0][i] = grid[0][i] + res[0][i-1];
        for(int j=1; j<m; j++)
            res[j][0] = grid[j][0] + res[j-1][0];

        for(int row=1; row<m; row++){
            for (int col=1; col<n; col++)
                res[row][col] = min(res[row-1][col] , res[row][col-1]) + grid[row][col];
        }
        return res[m-1][n-1];
    }
};

File: python-algorithm/algo_05_dynamic_plan/dp_7_matrix_path/L64-m.py
#--coding:utf-8--
'''
@Time   : 2020/10/12
@Author : No Name
'''

class Solution:
    def minPathSum(self, grid: List[List[int]]) -> int:
        # m n 别搞反了
        m, n = len(grid), len(grid[0])
        dp = [[0]*n for i in range(m)]
        dp[0][0] = grid[0][0]
        for j in range(1, n):
            dp[0][j] = dp[0][j-1]+grid[0][j]
        for i in range(1, m):
            dp[i][0] = dp[i-1][0]+grid[i][0]
        for i in range(1, m):
            for j in range(1, n):
                dp[i][j] = min(dp[i-1][j], dp[i][j-1]) + grid[i][j]
        return dp[m-1][n-1]